### PR TITLE
add ecrecover precompile

### PR DESCRIPTION
**Pull request by vgrabovski-lacero-platform**
signed-off-by: Vitali Grabovski <v.grabovski@lacero.io>

Related to https://github.com/hyperledger/burrow/issues/1240

### CODE REVIEW

File: execution/native/precompiles_test.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+	"github.com/hyperledger/burrow/crypto"
+
+	"github.com/hyperledger/burrow/execution/evm/asm/bc"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestSNativeContractDescription1a(t *testing.T) {
+	signatory := []byte("4fbdd7d412e5dad21d0c4add8b3cdc70ab65a5b3")
+	signature := []byte("e8cd78a90152a1549cfea3dcc8556e25618c67085c39e0b95e3d1ec6dbe9339525e7c61e005954fa67ce0c102cee434d1b111c4b0c52bad0fdc2663ca09451ac1b")
+	messageDigest := []byte("21400fb863b0ee7392af02491b6ca5fb87f68ad6af4867b104282dd8fe8b52e4")
```
Comment (silasdavis) [inline execution/native/precompiles_test.go:24]: Could we have a comment describing how these were generated - even better if we can generate them - we have little embedding utiltity that will compile solidity files and put their bytes and metadata into `.sol.go` files, see: https://github.com/monax/platform/blob/master/Makefile#L297-L299

File: execution/native/precompiles.go [hunk 2]
```diff
@@ -34,29 +39,33 @@ func leftPadAddress(bs ...byte) crypto.Address {
...
-	hash := input[:32]
-	v := byte(input[32] - 27) // ignore input[33:64], v is small.
-	sig := append(input[64:], v)
 
-	recovered, err := secp256k1.RecoverPubkey(hash, sig)
+	// SECP256K1 Recovery
+	hash := ctx.Input[:32] // from 0 to 31 bytes
+	sig := ctx.Input[63:]  // skip values in range[32:62]
+
+	publicKey, isCompressed, err := ecdsa.RecoverCompact(sig, hash)
```
Comment (silasdavis) [inline execution/native/precompiles.go:55]: Please our existing secp256k1-providing dependency `"github.com/btcsuite/btcd/btcec"`: `btcec.RecoverCompact` should do the job.

File: go.mod [hunk 3]
```diff
@@ -9,6 +9,7 @@ require (
 	github.com/asaskevich/govalidator v0.0.0-20180720115003-f9ffefc3facf // indirect
 	github.com/btcsuite/btcd v0.0.0-20190523000118-16327141da8c
 	github.com/cep21/xdgbasedir v0.0.0-20170329171747-21470bfc93b9
+	github.com/decred/dcrd/dcrec/secp256k1/v3 v3.0.0
```
Comment (silasdavis) [inline go.mod:12]: Looks like this is a fork of btcsuite secp256k1 - I would like avoid pulling in any new crypto dependencies that would need to be audited.

**Review by silasdavis — CHANGES_REQUESTED** [review]
Thanks very much, see comments inline if you can respond on thsoe I'd love to get this merged in. 

Can you rebase? I have (finally) fixed the CI

**Comment by silasdavis** [comment]
I've rolled this into #1420

**Comment by vgrabovski-lacero-platform** [comment]
Would like to work on your notes but it seems like you have already done it by yourself @silasdavis

**Comment by silasdavis** [comment]
Ah hello! Sorry I didn't mean to take if off you if you were still interested in finishing it off. I did base more work off your commit.

When looking into it I realised a few other things were not being done correctly with our handling of secp256k1 outside of ecrecover.

I don't know if you are at all interested in implementing some of the other elliptic curve precompiles. There is an abandoned PR with some of the work here: https://github.com/hyperledger/burrow/pull/1284

If that has any value or interest to you I'd be very pleased to assist with merging a PR

**Comment by vgrabovski-lacero-platform** [comment]
I don't mind. That is nice that better Burrow version in now in master! Well done

**Comment by vgrabovski-lacero-platform** [comment]
I will take a look at https://github.com/hyperledger/burrow/pull/1284 This mab be close to feature I would like to be included in Hyperledger_Burrow

**Pull request closed without merging**